<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>erl_error (erl_interface) -  (Erlang Documentation)</title>
    <link href="../erldocs.css" type="text/css" rel="stylesheet"/>
    <link href="/search.xml" rel="search" type="application/opensearchdescription+xml" title="erldocs"/>

    <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
      ga('create', 'UA-44246018-1', 'auto');
      ga('send', 'pageview');
    </script>
  </head>

  <body>
    <div id="sidebar" class="inactive">
      <input type="text" id="search" autocomplete="off" placeholder="press TAB to search"/>
      <ul id="results"> </ul>
    </div>

    <div id="content">
      <div style="margin:0px; padding:10px 20px;">
        
  
  <h1>erl_error</h1>
  <libsummary>Error Print Routines</libsummary>
  <div class="description">
    <p>This module contains some error printing routines taken
      from <em>Advanced Programming in the UNIX Environment</em> 
      by W. Richard Stevens. </p>
    <p>These functions are all called in the same manner as
      <code>printf()</code>, i.e. with a string containing format specifiers
      followed by a list of corresponding arguments. All output from
      these functions is to <code>stderr</code>.</p>
  </div>
  <div id="functions" class="category"><h4><a href="#functions">Functions</a></h4><hr />
    <div class="function">
      <pre>void erl_err_msg(FormatStr, ... )</pre>
      
      <ul class="type">
        <li><code>const char *FormatStr;</code></li>
      </ul>
      <div class="description">
        <p>The message provided by the caller is printed. This
          function is simply a wrapper for <code>fprintf()</code>.</p>
      </div>
    </div>
    <div class="function">
      <pre>void erl_err_quit(FormatStr, ... )</pre>
      
      <ul class="type">
        <li><code>const char *FormatStr;</code></li>
      </ul>
      <div class="description">
        <p>Use this function when a fatal error has occurred that
          is not due to a system call. The message provided by the
          caller is printed and the process terminates with an exit
          value of 1. The function does not return.</p>
      </div>
    </div>
    <div class="function">
      <pre>void erl_err_ret(FormatStr, ... )</pre>
      
      <ul class="type">
        <li><code>const char *FormatStr;</code></li>
      </ul>
      <div class="description">
        <p>Use this function after a failed system call. The message
          provided by the caller is printed followed by a string
          describing the reason for failure. </p>
      </div>
    </div>
    <div class="function">
      <pre>void erl_err_sys(FormatStr, ... )</pre>
      
      <ul class="type">
        <li><code>const char *FormatStr;</code></li>
      </ul>
      <div class="description">
        <p>Use this function after a failed system call. The message
          provided by the caller is printed followed by a string
          describing the reason for failure, and the process
          terminates with an exit value of 1. The function does not
          return.</p>
      </div>
    </div>
  </div>

  <div class="section">
    <h4>Error Reporting</h4>
    <p>Most functions in erl_interface report failures to the caller by
      returning some otherwise meaningless value (typically <code>NULL</code>
      or a negative number). As this only tells you that things did not
      go well, you will have to examine the error code in
      <code>erl_errno</code> if you want to find out more about the failure.</p>
  </div>
  <div id="functions" class="category"><h4><a href="#functions">Functions</a></h4><hr />
    <div class="function">
      <pre>volatile int erl_errno</pre>
      
      <div class="description">
        <p><code>erl_errno</code> is initially (at program startup) zero and
          is then set by many erl_interface functions on failure to a
          non-zero error code to indicate what kind of error it
          encountered. A successful function call might change
          <code>erl_errno</code> (by calling some other function that
          fails), but no function will ever set it to zero. This means
          that you cannot use <code>erl_errno</code> to see <em>if</em> a
          function call failed. Instead, each function reports failure
          in its own way (usually by returning a negative number or
          <code>NULL</code>), in which case you can examine <code>erl_errno</code>
          for details.</p>
        <p><code>erl_errno</code> uses the error codes defined in your
          system's <code>&lt;errno.h&gt;</code>.</p>
        <div class="note"><h2>Note!</h2>
          <p>Actually, <code>erl_errno</code> is a "modifiable lvalue" (just
            like ISO C defines <code>errno</code> to be) rather than a
            variable. This means it might be implemented as a macro
            (expanding to, e.g., <code>*_erl_errno()</code>). For reasons of
            thread- (or task-)safety, this is exactly what we do on most
            platforms.</p>
        </div>
      </div>
    </div>
  </div>

      </div>
      <!-- <div id="funwrapper"> -->
      <!--   <a id="viewfuns">View Functions</a> -->
      <!--    -->
      <!-- </div> -->
    </div>

    <script type="text/javascript">
      var CURRENT_ROOT = "../";
    </script>

    <script type="text/javascript" src="../jquery.js"></script>
    <script type="text/javascript" src="../erldocs_index.js"></script>
    <script type="text/javascript" src="../erldocs.js"></script>
  </body>
</html>
